### PR TITLE
Fix sending issues

### PR DESCRIPTION
**Pull request by pfista**
Added some changes @khamidou

### CODE REVIEW

File: spec/inbox_spec.rb [hunk 1]
```diff
@@ -118,7 +118,7 @@
 
       context "when the server responds with JSON that does not represent an array" do
         it "should raise an UnexpectedResponse" do
-          allow(@result).to receive(:code).and_return(500)
+          allow(@result).to receive(:code).and_return(200)
           expect {
             Inbox.interpret_response(@result, "{\"_id\":\"5107089add02dcaecc000003\",\"created_at\":\"2013-01-28T23:24:10Z\",\"domain\":\"generic\",\"name\":\"Untitled\",\"password\":null,\"slug\":\"\",\"tracers\":[{\"_id\":\"5109b5e0dd02dc5976000001\",\"created_at\":\"2013-01-31T00:08:00Z\",\"name\":\"Facebook\"},{\"_id\":\"5109b5f5dd02dc4c43000002\",\"created_at\":\"2013-01-31T00:08:21Z\",\"name\":\"Twitter\"}],\"published_pop_url\":\"http://group3.lvh.me\",\"unpopulated_api_tags\":[],\"unpopulated_api_regions\":[],\"label_names\":[]}", {:expected_class => Array})
```
Comment (khamidou) [inline spec/inbox_spec.rb]: Could you make this a little more readable? What's the shortest invalid JSON you can send here?

File: LICENSE.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
 The MIT License (MIT)
 ----
 
-Copyright (c) 2014 InboxApp, Inc. and Contributors
+Copyright (c) 2016 Nylas, Inc. and Contributors
```
Comment (khamidou) [inline LICENSE.txt]: lol

File: lib/inbox.rb [hunk 3]
```diff
@@ -22,59 +22,69 @@
...
+    attr_accessor :type
+    attr_accessor :message
     attr_accessor :server_error
 
-    def initialize(type, error, server_error = nil)
-      super(error)
-      self.error_type = type
+    def initialize(type, message, server_error = nil)
+      super(message)
+      self.type = type
```
Comment (khamidou) [inline lib/inbox.rb]: Curious, why rename this field?

File: lib/inbox.rb [hunk 4]
```diff
@@ -22,59 +22,69 @@
...
 
-    # Handle content expectation errors
+    # We expected a certain kind of object, but the API didn't return anything
     raise UnexpectedResponse.new if options[:expected_class] && result_content.empty?
-    json = options[:result_parsed]? result_content : JSON.parse(result_content)
-    if json.is_a?(Hash) && (json['type'] == 'api_error' or json['type'] == 'invalid_request_error')
+
+    # If it's already parsed, or if we've received an actual raw payload on success, don't parse
+    if options[:result_parsed] || (options[:raw_response] && result.code.to_i == 200)
+      response = result_content
```
Comment (khamidou) [inline lib/inbox.rb]: I think you can just `return result_content` here. No need to have this Pascal-style single-exit thing.

File: lib/inbox.rb [hunk 5]
```diff
@@ -22,59 +22,69 @@
...
+    else
+      response = JSON.parse(result_content)
```
Comment (khamidou) [inline lib/inbox.rb]: Same here.

File: lib/inbox.rb [hunk 6]
```diff
@@ -86,7 +96,7 @@ class API
     attr_accessor :api_server
     attr_accessor :stream_handler
     attr_reader :access_token
-    attr_reader :app_id
+    attr_reader :app_id
```
Comment (khamidou) [inline lib/inbox.rb]: Did you mean to have this here?

File: spec/draft_spec.rb [hunk 7]
```diff
@@ -57,29 +57,41 @@
...
-      draft.to = [{:name => 'Helena Handbasket', :email => 'helena@nylas.com'}]
-      expect(draft.id).to be nil
-
-      expect { draft.send! }.to raise_error(Inbox::InvalidRequest)
-      begin
-        draft.send!
-      rescue Inbox::InvalidRequest => e
-        expect(e.message).to eq("Invalid recipient address benbitdiddle@gmailcom")
-        expect(e.server_error).to eq("SPAM")
+    error_codes = [[400, Inbox::InvalidRequest],
```
Comment (khamidou) [inline spec/draft_spec.rb]: Hmm, I would just create an array from `HTTP_CODE_TO_EXCEPTIONS`. No need to duplicate the array here.

File: lib/inbox.rb (hunk 4 above)
Comment (pfista) [inline lib/inbox.rb]: I use response later in the method, so that's why I didn't return early. See L86

File: lib/inbox.rb (hunk 6 above)
Comment (pfista) [inline lib/inbox.rb]: Nope!

File: spec/draft_spec.rb [hunk 8]
```diff
@@ -57,16 +57,7 @@
...
-                   [402, Inbox::MessageRejected],
-                   [403, Inbox::AccessDenied],
-                   [404, Inbox::ResourceNotFound],
-                   [422, Inbox::MailProviderError],
-                   [429, Inbox::SendingQuotaExceeded],
-                   [500, Inbox::InternalError],
-                   [502, Inbox::BadGateway],
-                   [503, Inbox::ServiceUnavailable]]
-
+    error_codes = Inbox::HTTP_CODE_TO_EXCEPTIONS.to_a
```
Comment (khamidou) [inline spec/draft_spec.rb]: 👍

**Pull request merged**